src/numpy-stubs/_typing/_ufunc.pyi

```diff
@@ -380,6 +446,113 @@ class _Call21Bool(Protocol):
         **kwds: Unpack[_Kwargs3],
     ) -> np.bool | NDArray[np.bool]: ...
 
+@type_check_only
+class _Call21Logical(Protocol):
+    @overload  # (scalar, scalar, dtype: np.object_) -> np.object_
+    def __call__(
+        self,
+        x1: _ScalarLike_co,
+        x2: _ScalarLike_co,
+        /,
+        dtype: _DTypeLike[np.object_],
+        out: None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> bool: ...
+    @overload  # (scalar, scalar) -> bool
+    def __call__(
+        self,
+        x1: _NumberLike_co,
+        x2: _NumberLike_co,
+        /,
+        out: None = None,
+        dtype: _DTypeLikeBool | None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> np.bool: ...
+    @overload  # (array-like, array, dtype: object_) -> Array[object_]
+    def __call__(
+        self,
+        x1: _ArrayLikeNumber_co | _ArrayLikeObject_co,
+        x2: _AnyArray,
+        /,
+        dtype: _DTypeLike[np.object_],
+        out: None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> NDArray[np.object_]: ...
+    @overload  # (array-like, array, dtype: dtype[T]) -> Array[T]
+    def __call__(
+        self,
+        x1: _ArrayLikeNumber_co,
+        x2: NDArray[np.bool | np.number],
+        /,
+        out: None = None,
+        dtype: _DTypeLikeBool | None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> NDArray[np.bool]: ...
+    @overload  # (array, array-like, dtype: object_) -> Array[object_]
+    def __call__(
+        self,
+        x1: _AnyArray,
+        x2: _ArrayLikeNumber_co | _ArrayLikeObject_co,
+        /,
+        dtype: _DTypeLike[np.object_],
+        out: None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> NDArray[np.object_]: ...
+    @overload  # (array, array-like, dtype: dtype[T]) -> Array[T]
+    def __call__(
+        self,
+        x1: NDArray[np.bool | np.number] | _NestedSequence[np.bool | np.number],
+        x2: _ArrayLikeNumber_co,
+        /,
+        out: None = None,
+        dtype: DTypeLike | None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> NDArray[np.bool]: ...
+    @overload  # (array-like, array-like, out: T) -> T
+    def __call__(
+        self,
+        x1: _ArrayLikeNumber_co,
+        x2: _ArrayLikeNumber_co,
+        /,
+        out: _Out1[_ArrayT],
+        *,
+        dtype: None = None,
+        **kwds: Unpack[_Kwargs3],
+    ) -> _ArrayT: ...
+    @overload  # (array-like, array) -> Array[?]
+    def __call__(
+        self,
+        x1: _ArrayLikeNumber_co,
+        x2: _AnyArray,
```

Comment:
The `_AnyArray` here causes `(x1: NDArray[float64], x2: NDArray[np.object_])` to result in `NDArray[np.bool]`, even though it would be `NDArray[np.object_]` at runtime. Luckily, `NDArray[object_]` is easy to exclude:

```suggestion
        x1: _ArrayLikeNumber_co,
        x2: NDArray[np.bool | np.number] | _NestedSequence[np.bool | np.number | complex],
```

(I also took the liberty of adding a cherry on top, hopefully you don't mind)